net: Restrict period when cs_vNodes mutex is locked

hebasto opened the pull request:
This PR restricts the period when the `cs_vNodes` mutex is locked, prevents the only case when `cs_vNodes` could be locked before the `::cs_main`.

This change makes the explicit locking of recursive mutexes in the explicit order redundant.

jnewbery commented:
This does indeed seem useless. The mutex was added by @TheBlueMatt in commit d7c58ad514ee00db00589216166808258bc16b60:

>    Split CNode::cs_vSend: message processing and message sending
>    
>    cs_vSend is used for two purposes - to lock the datastructures used
>    to queue messages to place on the wire and to only call
>    SendMessages once at a time per-node. I believe SendMessages used
>    to access some of the vSendMsg stuff, but it doesn't anymore, so
>    these locks do not need to be on the same mutex, and also make
>    deadlocking much more likely.

We're implicitly guaranteed to only call `SendMessages()` serially since it's only ever called by message handler thread. If we want to be explicit about it, it'd be better to add a global lock to `PeerManagerImpl` (or per-`Peer` object), so that net_processing is enforcing its own synchronization internally.

hebasto commented:
Rebased 8ca2ee6377f534b15b47440b819b1564c9c06b88 -> b58097facb4fec80df8fa140dfefd9c461c93832 ([pr21563.02](https://github.com/hebasto/bitcoin/commits/pr21563.02) -> [pr21563.03](https://github.com/hebasto/bitcoin/commits/pr21563.03)) due to the conflict with #21571.

jnewbery commented:
This seems safe to me. The only thing that `CNode.cs_sendProcessing` enforces is preventing `SendMessages()` from being called for the same `CNode` concurrently. However, `SendMessages()` can only be called by `threadMessageHandler`, so there's no possibility of it being called concurrently.

I wonder if instead of simply removing this, we should replace it with something that more closely documents our expectations. I think that calling any of the `NetEventsInterface` methods (`InitializeNode`, `FinalizeNode`, `SendMessages`, `ProcessMessages`) in `PeerManagerImpl` concurrently could lead to problems, so perhaps a global mutex in `PeerManagerImpl` should be added that's taken whenever any of those functions are called? If we ever want to add concurrency to `PeerManagerImpl`, we could look at loosening that restriction.

hebasto commented:
@jnewbery 
> I wonder if instead of simply removing this, we should replace it with something that more closely documents our expectations. I think that calling any of the `NetEventsInterface` methods (`InitializeNode`, `FinalizeNode`, `SendMessages`, `ProcessMessages`) in `PeerManagerImpl` concurrently could lead to problems, so perhaps a global mutex in `PeerManagerImpl` should be added that's taken whenever any of those functions are called? If we ever want to add concurrency to `PeerManagerImpl`, we could look at loosening that restriction.

Maybe postpone it until #19398 is fulfilled?

jnewbery commented:
> Maybe postpone it until #19398 is fulfilled?

I think they can be done independently. But I think I'd like to see the new internal PeerManagerImpl lock being introduced in the same PR that removes cs_sendProcessing.

hebasto committed:
net: Combine two loops into one, and update comments

hebasto committed:
net: Restrict period when cs_vNodes mutex is locked

hebasto commented:
Updated b58097facb4fec80df8fa140dfefd9c461c93832 -> 4de76058f8c3150de07d7a02faaceb261d732499 ([pr21563.03](https://github.com/hebasto/bitcoin/commits/pr21563.03) -> [pr21563.04](https://github.com/hebasto/bitcoin/commits/pr21563.04)):

- rebased on top of the recent changes in CI
- addressed @jnewbery's [comment](https://github.com/bitcoin/bitcoin/pull/21563#issuecomment-816538311)

MarcoFalke commented:
```
 node0 stderr libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument 
```

https://cirrus-ci.com/task/5038649289998336?logs=ci#L3615

hebasto commented:
> ```
>  node0 stderr libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument 
> ```
> 
> https://cirrus-ci.com/task/5038649289998336?logs=ci#L3615

Looks like a mutex is prematurely destroyed. Weird.

hebasto commented:
Updated 4de76058f8c3150de07d7a02faaceb261d732499 -> b1e5ca217c3557ed89d985c7e8ea28c2b9ed303b ([pr21563.04](https://github.com/hebasto/bitcoin/commits/pr21563.04) -> [pr21563.05](https://github.com/hebasto/bitcoin/commits/pr21563.05), [diff](https://github.com/hebasto/bitcoin/compare/pr21563.04..pr21563.05)):

- fixed bug

jnewbery commented on src/init.cpp:
```diff
@@ -200,18 +200,8 @@ void Shutdown(NodeContext& node)
-    // Thus the implicit locking order requirement is: (1) cs_main, (2) g_cs_orphans, (3) cs_vNodes.
     if (node.connman) {
         node.connman->StopThreads();
-        LOCK2(::cs_main, ::g_cs_orphans);
         node.connman->StopNodes();
```
This can now be replaced with `if (node.connman) node.connman->Stop()`.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/21563#issuecomment-818709445).

jnewbery commented on src/net.cpp:
```diff
@@ -2636,6 +2633,9 @@ void CConnman::StopNodes()
+    std::vector<CNode*> nodes;
+    std::list<CNode*> nodes_disconnected;
+    {
     LOCK(cs_vNodes);
```
I think this lock can be held for much less time:

```c++
    // Delete peer connections
    std::vector<CNode*> nodes;
    {
        LOCK(cs_vNodes);
        nodes = std::move(vNodes);
        vNodes.clear();
    }

    for (CNode* pnode : nodes) {
        pnode->CloseSocketDisconnect();
        DeleteNode(pnode);
    }

    for (CNode* pnode : vNodesDisconnected) {
        DeleteNode(pnode);
    }
    vNodesDisconnected.clear();

    // Close listening sockets
    for (ListenSocket& hListenSocket : vhListenSocket) {
        if (hListenSocket.socket != INVALID_SOCKET) {
            if (!CloseSocket(hListenSocket.socket)) {
                LogPrintf("CloseSocket(hListenSocket) failed with error %s\n", NetworkErrorString(WSAGetLastError()));
            }
        }
    }
    vhListenSocket.clear();

    semOutbound.reset();
    semAddnode.reset();
```

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/21563#issuecomment-818709445).

jnewbery commented on src/net_processing.cpp:
```diff
@@ -238,10 +238,10 @@ class PeerManagerImpl final : public PeerManager
-    void FinalizeNode(const CNode& node) override;
-    bool ProcessMessages(CNode* pfrom, std::atomic<bool>& interrupt) override;
-    bool SendMessages(CNode* pto) override EXCLUSIVE_LOCKS_REQUIRED(pto->cs_sendProcessing);
+    void InitializeNode(CNode* pnode) override EXCLUSIVE_LOCKS_REQUIRED(!m_net_events_mutex);
```
What do you think about excluding holding `cs_main` when calling into a NetEventsInterface method?

hebasto replied:
It is not trivial due to the https://github.com/bitcoin/bitcoin/blob/c1f480fb234856a60e9b1f4bf97114bfd9b3d0b0/src/net_processing.cpp#L662

Going to keep the scope of this PR tide.

jnewbery commented on src/init.cpp:
```diff
@@ -200,18 +200,8 @@ void Shutdown(NodeContext& node)
-    // Thus the implicit locking order requirement is: (1) cs_main, (2) g_cs_orphans, (3) cs_vNodes.
     if (node.connman) {
         node.connman->StopThreads();
-        LOCK2(::cs_main, ::g_cs_orphans);
```
You can also remove the `LOCK2(::cs_main, g_cs_orphans);` from `fuzz/process_message.cpp` and `fuzz/process_messages.cpp`.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/21563#issuecomment-818709445).

hebasto commented:
Updated b1e5ca217c3557ed89d985c7e8ea28c2b9ed303b -> a3663329ad1899cc012437167a316040d77e00bc ([pr21563.05](https://github.com/hebasto/bitcoin/commits/pr21563.05) -> [pr21563.06](https://github.com/hebasto/bitcoin/commits/pr21563.06)):

- rebased on top of the recent CI changes
- addressed @jnewbery's comments

jnewbery commented:
This seems good to me.

@MarcoFalke - you introduced the `LOCK2(::cs_main, ::g_cs_orphans);` in #18458. Do you see any problem with locking `cs_vNodes`, grabbing a copy of `vNodes`, releasing the lock and then deleting the nodes in `CConnman::Stop()`? This is similar to what happens in the socket handler and message handler threads?

ajtowns commented:
Removing a mutex that guards nothing then adding a mutex that also doesn't actually guard anything seems a bit backwards...

#21527 has it guarding the extratxns data structures (as part of getting rid of g_cs_orphans). It would be good for it to be able to guard the address data structures in struct Peer, but without it being a global, that will probably be awkward at best, since Peer doesn't have a reference back to PeerManagerImpl.

The cs_vNodes changes don't seem to be mentioned in the title or PR description?

MarcoFalke commented:
The fist commit seems separate from the other changes?

hebasto commented:
Updated a3663329ad1899cc012437167a316040d77e00bc -> 9766b7f43d4fedd1740f4444689b6ba2ff33d12d ([pr21563.06](https://github.com/hebasto/bitcoin/commits/pr21563.06) -> [pr21563.07](https://github.com/hebasto/bitcoin/commits/pr21563.07)):

- dropped all commits but the first one in favor of #21527

The PR description updated.

jnewbery commented:
I think this change is fine. It steals the `CNode*`s from `vNodes`, releases the mutex and then cleans up the nodes. That's very similar to the pattern in `SocketHandler()`:

https://github.com/bitcoin/bitcoin/blob/018045347110d76cfc1f89a79fb49a867e3b8449/src/net.cpp#L1481-L1487

and in `ThreadMessageHandler()`:

https://github.com/bitcoin/bitcoin/blob/018045347110d76cfc1f89a79fb49a867e3b8449/src/net.cpp#L2185-L2192

The difference being that in those cases, an extra reference is taken and `nRefCount` is incremented. Here, the reference count is not incremented since the pointer is stolen from `vNodes` before clearing that vector.

Perhaps we could document our assumptions by calling `Release()` on the `CNode` object when stealing it from `vNodes` or immediately before calling `DeleteNode()`, and then asserting that `nRefCount == 0` in the `CNode` destructor? Maybe in a follow up.

ajtowns commented on src/net.cpp:
```diff
@@ -2635,25 +2635,34 @@ void CConnman::StopNodes()
+    {
+        LOCK(cs_vNodes);
+        nodes = std::move(vNodes);
+        vNodes.clear();
```
Could write this as:

```c++
cs_vNodes.lock();
std::vector<CNode*> nodes(std::move(vNodes)); // move constructor clears vNodes
cs_vNodes.unlock();
``` 

If I'm reading the spec right, the move constructor is guaranteed to be constant time, while `operator=` is linear time; and the move constructor also guarantees the original ends up cleared. Since C++17 the move constructor is also marked noexcept, so `lock` and `unlock` in place of RAII locks should be sound. Alternatively, could use `WAIT_LOCK(cs_vNodes, lock); ...; REVERSE_LOCK(lock);`.

MarcoFalke replied:
Does performance matter here at all? Shutdown only happens once and flushing to disk will take magnitudes longer anyway.

ajtowns replied:
No I don't think so -- at worst it's just moving pointers around, and only a hundred or so of them in normal configurations.

jnewbery replied:
> the move constructor is guaranteed to be constant time, while operator= is linear time

This is the move assignment operator (number 2 in https://en.cppreference.com/w/cpp/container/vector/operator%3D). Its complexity is linear in the size of `this` - the vector being moved to. Presumably that's because the objects in `this` need to be destructed and the memory already owned by `this` needs to deallocated. In our case, `this` is empty, and so the operation is constant time - it just needs to copy start pointer/size/capacity from `other` to `this`.

> Does performance matter here at all?

Absolutely not.

> I wasn't sure if std::move + clear was safe / necessary

Moving from a vector leaves it in a "valid but unspecified state". The `clear()` is _probably_ unnecessary since this is shutdown and we're not going to touch `vNodes` again, but I think it's good practice to not to leave vectors in an unspecified state if we can help it.

ajtowns reviewed (commented):
Seems sensible at first glance, and I think any bugs would be caught by CI. Need to have a more thorough look though.

MarcoFalke commented on src/net.cpp:
```diff
@@ -2635,25 +2635,34 @@ void CConnman::StopNodes()
+    std::vector<CNode*> nodes;
+    {
+        LOCK(cs_vNodes);
+        nodes = std::move(vNodes);
+        vNodes.clear();
```
Could achieve the same in one less line of code with https://en.cppreference.com/w/cpp/container/vector/swap ?

MarcoFalke replied:
Or even remove all of this logic and re-use the existing "disconnect all nodes" logic?

```diff
diff --git a/src/net.cpp b/src/net.cpp
index b7c1b8c6c4..9a4a989e15 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -2636,31 +2636,17 @@ void CConnman::StopNodes()
     }
 
     // Delete peer connections
-    std::vector<CNode*> nodes;
     {
         LOCK(cs_vNodes);
-        nodes = std::move(vNodes);
-        vNodes.clear();
+        assert(vNodes.empty());
     }
-
-    for (CNode* pnode : nodes) {
-        pnode->CloseSocketDisconnect();
-        DeleteNode(pnode);
-    }
-
-    for (CNode* pnode : vNodesDisconnected) {
-        DeleteNode(pnode);
-    }
-    vNodesDisconnected.clear();
+    assert(vNodesDisconnected.empty());
 
     // Close listening sockets
-    for (ListenSocket& hListenSocket : vhListenSocket) {
-        if (hListenSocket.socket != INVALID_SOCKET) {
-            if (!CloseSocket(hListenSocket.socket)) {
+    for (ListenSocket& hListenSocket : vhListenSocket)
+        if (hListenSocket.socket != INVALID_SOCKET)
+            if (!CloseSocket(hListenSocket.socket))
                 LogPrintf("CloseSocket(hListenSocket) failed with error %s\n", NetworkErrorString(WSAGetLastError()));
-            }
-        }
-    }
     vhListenSocket.clear();
 
     semOutbound.reset();
diff --git a/src/net.h b/src/net.h
index 13f86b6a91..5a87a5a304 100644
--- a/src/net.h
+++ b/src/net.h
@@ -867,6 +867,8 @@ public:
     void Stop()
     {
         StopThreads();
+fNetworkActive=true;
+DisconnectNodes();
         StopNodes();
     };

MarcoFalke replied:
(I tried running the tests with my diff and for some reason the node crashed)

hebasto replied:
> Could achieve the same in one less line of code with https://en.cppreference.com/w/cpp/container/vector/swap ?

Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/21563#issuecomment-824898734).

MarcoFalke replied:
Just to answer (myself) why my diff crashed: When the threads are interrupted, the node refcounts do *not* get decremented

vasild replied:
I am working on a change that will address this problem of forgetting `Release()` due to an early `return`.

vasild replied:
> When the threads are interrupted, the node refcounts do _not_ get decremented

This is changed/fixed in https://github.com/bitcoin/bitcoin/pull/21878 in commit

96bcc72b5 `net: dedup and RAII-fy the creation of a copy of CConnman::vNodes`

MarcoFalke commented on src/net.cpp:
```diff
@@ -2635,25 +2635,34 @@ void CConnman::StopNodes()
+                LogPrintf("CloseSocket(hListenSocket) failed with error %s\n", NetworkErrorString(WSAGetLastError()));
+            }
+        }
+    }
```
minor nit: would be nice to do style-fixups in a separate commit

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/21563#issuecomment-824898734).

vasild reviewed (approved):
ACK 9766b7f43d4fedd1740f4444689b6ba2ff33d12d

Something to consider for further improvement:

It is not even necessary to lock `cs_vNodes` inside `StopNodes()`. Why? Because by the time `StopNodes()` is called the other threads have been shut down.

If it _was_ necessary to protect the code in `StopNodes()` with a mutex, then this PR would be broken:
```cpp
lock()
tmp = std::move(shared_vector)
unlock()
destroy each element in tmp
// What prevents new entries from being added to shared_vector
// here, since we have unlock()ed? Nothing. So when this function
// completes, shared_vector is still alive and kicking (with new
// elements being added to it in the meantime).
```

The `LOCK(cs_vNodes);` is only needed to keep the [thread safety analysis](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html) happy. This code actually belongs to the destructor `~CConnman()` where we can access `vNodes` without locking and without upsetting the TSA. I think it can be moved to the destructor with some further changes (outside of this PR).

jnewbery commented:
> This code actually belongs to the destructor ~CConnman() where we can access vNodes without locking and without upsetting the TSA. I think it can be moved to the destructor with some further changes (outside of this PR).

Maybe, but that's a much more invasive change. `CConnman::Stop()` calls `PeerManager::DeleteNode()` for all the nodes. The destructor for `CConnman` is called _after_ the destructor for `PeerManager` (in the reverse order that they're constructed). That can be changed, but we'd need be very careful.

jnewbery commented:
utACK 9766b7f43d4fedd1740f4444689b6ba2ff33d12d

Agree with @MarcoFalke that the move & clear is more elegantly expressed as std::vector::swap().

<details>
<summary>Diff</summary>

```diff
diff --git a/src/net.cpp b/src/net.cpp
index b7c1b8c6c4..94029491ab 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -2637,11 +2637,7 @@ void CConnman::StopNodes()
 
     // Delete peer connections
     std::vector<CNode*> nodes;
-    {
-        LOCK(cs_vNodes);
-        nodes = std::move(vNodes);
-        vNodes.clear();
-    }
+    WITH_LOCK(cs_vNodes, nodes.swap(vNodes));
 
     for (CNode* pnode : nodes) {
         pnode->CloseSocketDisconnect();
```

</details>

hebasto committed:
net, refactor: Fix style in CConnman::StopNodes

hebasto commented:
Updated 9766b7f43d4fedd1740f4444689b6ba2ff33d12d -> 8c8237a4a10feb2ac9ce46f67b5d14bf879b670f ([pr21563.07](https://github.com/hebasto/bitcoin/commits/pr21563.07) -> [pr21563.08](https://github.com/hebasto/bitcoin/commits/pr21563.08), [diff](https://github.com/hebasto/bitcoin/compare/pr21563.07..pr21563.08)):

- addressed @MarcoFalke's comments:
  - https://github.com/bitcoin/bitcoin/pull/21563#discussion_r617241012
> Could achieve the same in one less line of code with https://en.cppreference.com/w/cpp/container/vector/swap ?
  - https://github.com/bitcoin/bitcoin/pull/21563#discussion_r617259352
> minor nit: would be nice to do style-fixups in a separate commit

vasild reviewed (approved):
ACK 8c8237a4a10feb2ac9ce46f67b5d14bf879b670f

jnewbery commented:
utACK 8c8237a4a10feb2ac9ce46f67b5d14bf879b670f

ajtowns commented on src/net.cpp:
```diff
@@ -2636,22 +2636,22 @@ void CConnman::StopNodes()
-    vNodes.clear();
     vNodesDisconnected.clear();
     vhListenSocket.clear();
     semOutbound.reset();
```
Might be worth adding a comment as to why `vNodesDisconnected`, `vhListenSocket`, `semOutbound` and `semAddnode` are all safe to reset here -- I think it's because:

 * `vNodesDisconnected` and `vhListenSocket` are only otherwise accessed by `ThreadSocketHandler`
 * `semOutbound` and `semAddnode` are only otherwise accessed via `ThreadOpenConnections`, `Start()`, `Interrupt()` (and `Interrupt()` is only safe if it's invoked in between `Start()` and `StopNodes()`), and `AddConnection` (which is called from rpc/net.cpp so also requires that we won't be adding connections via RPC -- might be good to add an `if (semOutbound == nullptr) return false;` to `AddConnection()`)

MarcoFalke replied:
RPC must be (and is assumed to be) shutdown before `Stop` is called. The other threads are also assumed to be shut down. Maybe reverting my commit (https://github.com/bitcoin/bitcoin/pull/21563#issuecomment-826278304) could help to self-document that better?

ajtowns commented on src/test/fuzz/process_message.cpp:
```diff
@@ -100,7 +100,6 @@ void fuzz_target(FuzzBufferType buffer, const std::string& LIMIT_TO_MESSAGE_TYPE
     }
     SyncWithValidationInterfaceQueue();
-    LOCK2(::cs_main, g_cs_orphans); // See init.cpp for rationale for implicit locking order requirement
     g_setup->m_node.connman->StopNodes();
```
I think there probably should be a comment as to why explicitly calling `StopNodes` is still necessary

(I believe it's because if you don't do that, then `m_node` will get destructed, deleting its `peerman`, then attempting to delete it's `connman` which will see some entries in `vNodes` and try to call `peerman->FinalizeNode()` on them, but `peerman` is deleted at that point. With the lock order fixed, it may be possible to have `~NodeContext` call `if (peerman) { peerman->Stop(); }` an remove those lines from the tests entirely. Not sure if that would also let you remove it from init.cpp)

MarcoFalke replied:
I think changing the destruction order can be done in a separate pull?

ajtowns reviewed (commented):
utACK 8c8237a4a10feb2ac9ce46f67b5d14bf879b670f - logic seems sound

Couple of comment improvements that would be good; longer term would also be better to add missing guards to things in net so that the compiler can catch more errors, and simplify the ownership/interaction between CConnman/PeerManager/NodeContext...

vasild commented:
(consider out of scope of this PR)

Could we assert that all threads are stopped when `StopNodes()` starts executing? Something like
```cpp
assert(threadMessageHandler.get_id() == std::thread::id());
assert(threadSocketHandler.get_id() == std::thread::id());
...
```

Or even call `StopThreads()` at the start of `StopNodes()`?

MarcoFalke commented:
> Or even call StopThreads() at the start of StopNodes()?

This used to be the case before commit fa369651c5523f393e0bfcfa328b8e27006711aa. Maybe that commit should be reverted now?

maflcko reviewed (approved):
I think the changes are nice and can be merged


review ACK 8c8237a4a10feb2ac9ce46f67b5d14bf879b670f 👢

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 8c8237a4a10feb2ac9ce46f67b5d14bf879b670f 👢
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiYLgwAy9ZbgTGcc2V3HKj7fNh6KWOQ2DgAVsBLAmRP1D+443eCC3twNxELvtOe
Mq2gmwseUZK1oNKU0RnCHvg5bZOqcId1N1Sf6BEbJOmkGFPktaVLTz2TmQpLLE42
ZTMrqWEJreQ9DXqEDDKV222Zu4ucLmUFuNqs5MYj+moUTBdedFcbeF96cJK2hBc1
OgTJXL6MLoMAFeL4PKS6j+E2yHXOFuKHWE9WHPJ0wFKujTqlBPtntZxoWb7fby2A
ZHrA8wxjdNQbgovYRjToVe/bif2SNuFbfNwE2XYalQ50U8bY5n3hC2SUB/poGTaJ
GoSOWiv1jAUimiMUVyicP+JDfprAjPkB9hXKC7k4kvbluTpqYz4UGxhnBmKnU/qe
+gU1fPldReMkVGg5YjG3pg7Kd4izdq2YzyNMXdyIRHJaWn+gvsl4vQjPLUe7+WFL
DvX4dkpJ9fHF/DoHLsYi5/g2sG3G2G1Z4jrV/TFqAInAa0tuEqdWwYBomHi5QBQX
GzrRIkFr
=lXKU
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `2744498e52ef3957483818d251ab7a217f1deba6d21ad39bc26dd3c078c33d07  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401082744498e52ef3957483818d251ab7a217f1deba6d21ad39bc26dd3c078c33d07f010d8283008508fb338ffde04f8cdf1d6ec08fff0100819c453ab2e3821703f7aa40f977d0908f1046085235af0089fb2a89004cc27180083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01023472992b49008ecdb7a2ae520996ace08f120d3fc3eacb453235fbc4d20d1a746e0af552b1085c0929154c6d980b7f6ad5ece08f1046085157ef008178961a2ee7e7a420083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010737674bb979f5ea438a9c8119b125b0c08f1046085235af008f45114f6a80245b70083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010cf395c6b73f7c5b45e2da74742275de908f1046085235bf008ad864af9c7c7ec040083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented:
There is a related pr, which is waiting for review donations: #21750
